File: services/configd/csharp/src/CtlFlow.Configuration.Configd.Service/Kubernetes/DeriveNativeName.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;

namespace CtlFlow.Configuration.Configd.Service.Kubernetes;

internal static partial class KubernetesNames
{
    internal static ValueTask<string> DeriveNativeName(
        string domain,
        string prefix,
        string identifier,
        CancellationToken cancellation)
    {
        cancellation.ThrowIfCancellationRequested();
        var domainBytes = Encoding.ASCII.GetBytes(domain);
        var identifierBytes = Encoding.UTF8.GetBytes(identifier);
        var content = new byte[
            domainBytes.Length + 1 + 4 + identifierBytes.Length];
        domainBytes.CopyTo(content, 0);
        BinaryPrimitives.WriteUInt32BigEndian(
            content.AsSpan(domainBytes.Length + 1, 4),
            checked((uint)identifierBytes.Length));
        identifierBytes.CopyTo(content, domainBytes.Length + 5);
        var digest = SHA256.HashData(content);
        return ValueTask.FromResult(
            prefix + Convert.ToHexString(digest.AsSpan(0, 16))
                .ToLowerInvariant());
    }
}

File: services/configd/tests/support/kubernetes/derive-native-name.ts
import {
  createHash
} from "node:crypto";

export function deriveNativeName(
  domain: string,
  prefix: string,
  identifier: string
): string {
  const domainBytes = Buffer.from(domain, "ascii");
  const identifierBytes = Buffer.from(identifier, "utf8");
  const length = Buffer.alloc(4);
  length.writeUInt32BE(identifierBytes.length);
  const digest = createHash("sha256")
    .update(domainBytes)
    .update(Buffer.of(0))
    .update(length)
    .update(identifierBytes)
    .digest();
  return prefix + digest.subarray(0, 16).toString("hex");
}
